vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/MongoIterableStream.java

```diff
@@ -0,0 +1,186 @@
+  private Handler<Throwable> exceptionHandler;
+  private Handler<Void> endHandler;
+  private boolean paused;
+  private boolean readInProgress;
+  private boolean closed;
+
+  MongoIterableStream(Context context, MongoIterable<JsonObject> mongoIterable) {
+    this.context = context;
+    this.mongoIterable = mongoIterable;
+  }
+
+  @Override
+  public synchronized MongoIterableStream exceptionHandler(Handler<Throwable> handler) {
+    checkClosed();
+    this.exceptionHandler = handler;
+    return this;
+  }
+
+  private void checkClosed() {
+    if (closed) {
+      throw new IllegalArgumentException("Stream is closed");
+    }
+  }
+
+  @Override
+  public synchronized MongoIterableStream handler(Handler<JsonObject> handler) {
+    checkClosed();
+    if (handler == null) {
+      close();
+    } else {
+      dataHandler = handler;
+      SingleResultCallback<AsyncBatchCursor<JsonObject>> callback = (result, t) -> {
+        context.runOnContext(v -> {
+          synchronized (this) {
+            if (t != null) {
+              close();
+              handleException(t);
+            } else {
+              batchCursor = result;
+              if (canRead()) {
+                doRead();
+              }
+            }
+          }
+        });
+      };
+      try {
+        mongoIterable.batchCursor(callback);
+      } catch (Exception e) {
+        close();
+        handleException(e);
+      }
+    }
+    return this;
+  }
+
+  private boolean canRead() {
+    return !paused && !closed;
+  }
+
+  @Override
+  public synchronized MongoIterableStream pause() {
+    checkClosed();
+    paused = true;
+    return this;
+  }
+
+  @Override
+  public synchronized MongoIterableStream resume() {
+    checkClosed();
+    if (paused) {
+      paused = false;
+      if (dataHandler != null) {
+        doRead();
+      }
+    }
+    return this;
+  }
+
+  private synchronized void doRead() {
+    if (readInProgress) {
+      return;
+    }
+    readInProgress = true;
+    if (queue == null) {
+      queue = new ArrayDeque<>(BATCH_SIZE);
+    }
+    if (!queue.isEmpty()) {
+      context.runOnContext(v -> emitQueued());
+      return;
+    }
+    batchCursor.setBatchSize(BATCH_SIZE);
+    context.<List<JsonObject>>executeBlocking(fut -> {
+      batchCursor.next((result, t) -> {
+        if (t != null) {
+          fut.fail(t);
+        } else {
+          fut.complete(result == null ? Collections.emptyList() : result);
+        }
+      });
+    }, false, ar -> {
+      synchronized (this) {
+        if (ar.succeeded()) {
+          queue.addAll(ar.result());
+          if (queue.isEmpty()) {
+            close();
+            if (endHandler != null) {
+              endHandler.handle(null);
+            }
+          } else {
+            emitQueued();
+          }
+        } else {
+          close();
+          handleException(ar.cause());
+        }
+      }
+    });
+  }
+
+  private void handleException(Throwable cause) {
```

Comment:
should it be synchronized ? The exception handler can be updated.

Reply:
`handleException` is always called from a synchronized method or block. The exception handler is guarded by the stream instance